src/resources/uart_esp32.cc

```diff
@@ -17,95 +17,584 @@
+  vRingbufferReturnItemFromISR(ring_buffer(), buffer, null);
+}
+
+void TxBuffer::write(const uint8_t* buffer, uint16_t length, uint8_t break_length) {
+  {
+    SpinLocker locker(&spinlock_);
 
-      case UART_BREAK:
-        // Ignore.
-        break;
+    TxTransferHeader header = {
+        .break_length_ = static_cast<uint8_t>(break_length),
+        .remaining_data_length_ = static_cast<uint16_t>(length)
+    };
 
-      default:
-        state |= kErrorState;
-        break;
+    if (xRingbufferSend(ring_buffer(), &header, sizeof(TxTransferHeader), 0) == pdFALSE) {
+      abort();
     }
 
-    return state;
+    if (xRingbufferSend(ring_buffer(), buffer, length, 0) == pdFALSE) {
+      abort();
+    }
   }
-};
 
-MODULE_IMPLEMENTATION(uart, MODULE_UART);
+  uart()->enable_interrupt_index(UART_TOIT_INTR_TXFIFO_EMPTY);
+}
+
+UART_ISR_INLINE uint8_t TxBuffer::read_break() {
+  IsrSpinLocker locker(&spinlock_);
+
+  uint8_t break_length = transfer_header_.break_length_;
+  transfer_header_.break_length_ = 0; // To avoid it being read again and the break continuing...
+  return break_length;
+}
+
+UART_ISR_INLINE uint8_t* TxBuffer::read(size_t *received, size_t max_length) {
+  IsrSpinLocker locker(&spinlock_);
+
+  if (transfer_header_.remaining_data_length_ == 0 && transfer_header_.break_length_ == 0) {
+    // No more data in the latest package. Need to read header
+    size_t header_received;
+    size_t read = 0;
+    do { // As the header is multiple bytes, it could be split over the edge of the ring buffer
+      void* header_data = xRingbufferReceiveUpToFromISR(ring_buffer(), &header_received, sizeof(TxTransferHeader) - read);
+      if (!header_data) {
+        // TxBuffer is empty
+        return null;
+      }
+      memcpy(reinterpret_cast<uint8_t*>(&transfer_header_) + read, header_data, header_received);
+      // No blocking calls ever on TxBuffer, so ignore last parameter
+      vRingbufferReturnItemFromISR(ring_buffer(), header_data, null);
+      read += header_received;
+    } while (read < sizeof(TxTransferHeader));
+  }
+
+  size_t maximum_to_receive = max_length;
+  if (transfer_header_.remaining_data_length_ < maximum_to_receive) maximum_to_receive = transfer_header_.remaining_data_length_;
+
+  if (maximum_to_receive == 0) {
+    return null;
+  }
+
+  auto buffer = static_cast<uint8_t*>(xRingbufferReceiveUpToFromISR(ring_buffer(), received, maximum_to_receive));
+  transfer_header_.remaining_data_length_ -= *received;
+  return buffer;
+}
+
+size_t TxBuffer::free_size() {
+  size_t ringbuffer_free = xRingbufferGetCurFreeSize(ring_buffer());
+  if (ringbuffer_free < sizeof(TxTransferHeader)) return 0;
+  return ringbuffer_free - sizeof(TxTransferHeader);
+}
+
+void UART_ISR_INLINE RxBuffer::send(uint8_t* buffer, uint32_t length) {
+  // We are never blocking on the ring buffer
+  xRingbufferSendFromISR(ring_buffer(), buffer, length, null);
+}
+
+void RxBuffer::read(uint8_t* buffer, uint32_t length) {
```

Comment:
Do I understand it right that we only call this if there is enough data in the ring buffer?  So the reason for the loop is that there might be a wraparound, so that two reads are needed.  Please add a comment that says something like this (if it's correct).